Remove null, blank from start field in Schedules. Step 4/5

hunytalk opened the pull request:
Removes null & blank attributes from start_date field in schedules model in column rename release also remove writes from old field where possible (all writes to old field field 'start' will be removed when dropping field). Step 4/5
DE-1825

1. Add new column (https://github.com/edx/edx-platform/pull/22372/ )
2. Add writes to go in new column ( https://github.com/edx/edx-platform/pull/22373 )
3. Data migration from old to new column
  - Custom management command for data migration (https://github.com/edx/edx-platform/pull/22496 )
- Data migration PR. ( https://github.com/edx/edx-platform/pull/22391 )
4. Remove Null, blank from start_date and remove writes from old field ( **Current PR** )
5. Remove old field and also writes from old field. ( https://github.com/edx/edx-platform/pull/22392/ )

### Please consider the following when opening a pull request:

- Link to the relevant JIRA ticket(s) and tag any relevant team(s).
- Squash your changes down into one or more discrete commits.
  In each commit, include description that could help a developer
  several months from now.
- If running `make upgrade`, run _as close to the time of merging as possible_
  to avoid accidentally downgrading someone else's package.
  Put the output of `make upgrade` in its own separate commit,
  decoupled from other code changes.
- Aim for comprehensive test coverage, but remember that
  automated testing isn't a substitute for manual verification.
- Carefully consider naming, code organization, dependencies when adding new code.
  Code that is amenable to refactoring and improvement benefits all platform developers,
  especially given the size and scope of edx-platform.
  Consult existing Architectural Decision Records (ADRs),
  including those concerning the app(s) you are changing and
  [those concerning edx-platform as a whole](https://github.com/edx/edx-platform/tree/master/docs/decisions).

macdiesel commented on common/djangoapps/student/tests/test_views.py:
```diff
@@ -777,7 +777,11 @@ def test_content_gating_course_card_changes(self):
         )
-        schedule = ScheduleFactory(start=self.THREE_YEARS_AGO + timedelta(days=1), enrollment=enrollment)
+        schedule = ScheduleFactory(
+            start=self.THREE_YEARS_AGO + timedelta(days=1),
```
Unlikely but what if somehow the execution of this spans 2 days? We should calculate the delta once and use that variable for the of the ScheduleFactory.

hunytalk replied:
Ok, will change this to use variable. 
I didn't change it as this change was temporary since 'start' will be removed in next release.

doctoryes commented:
Two general notes:
- it'd be nice to have links in this PR back to the previous three PRs which were merged before this one was reviewed.
- I'm noticing a mix of adding writes to `start_date`, removing `start` sometimes, and keeping `start` other times. Shouldn't the writing to `start_date` have happened when the field was added? Why can't writes to `start` be removed in all cases in this PR?

doctoryes commented:
A PR message of this format is what I've tried to use:
```
https://openedx.atlassian.net/browse/DE-1826

This stage does the following:
- Adds the new field and migration to create the column.
- Makes all writes go to both old and new field.

Stage 1: this PR.
Stage 2: #22709 
Stage 3: #22753 

Testing plan detailed in https://openedx.atlassian.net/wiki/spaces/DE/pages/1065681172/Patterns+For+Renaming+Django+Model+Fields#PatternsForRenamingDjangoModelFields-RolloutPlanForCreditRequirement.orderIntheLMS
```

hunytalk commented:
> Two general notes:
> 
>     * it'd be nice to have links in this PR back to the previous three PRs which were merged before this one was reviewed.
> 
>     * I'm noticing a mix of adding writes to `start_date`, removing `start` sometimes, and keeping `start` other times. Shouldn't the writing to `start_date` have happened when the field was added? Why can't writes to `start` be removed in all cases in this PR?

I've added previous steps as you mentioned.
Yes it should but mistakenly some tests have missed adding writes to new field (`start_date`) when i created PR for 'adding writes to new field', that's why i've added them here. I've tested it in devstack locally and is working fine.
All writes to old field `start` could not be removed here as it's a non-NULL & blank field.

macdiesel commented on openedx/core/djangoapps/schedules/models.py:
```diff
@@ -24,10 +24,7 @@ class Schedule(TimeStampedModel):
-        # TODO remove blank=True and null=True once data migration is complete in column rename release.
-        blank=True,
         db_index=True,
-        null=True,
```
This line should have been the clue to tip us off that an AlterField would be run against this table.

hunytalk replied:
Yes unfortunately, it is.

macdiesel commented on openedx/core/djangoapps/schedules/migrations/0010_remove_null_blank_from_schedules_date.py:
```diff
@@ -0,0 +1,20 @@
+    ]
+
+    operations = [
+        migrations.AlterField(
```
I believe this line should also have tipped us off that an alter statement we would not want to run would be executed.
